pkg/atomicwriter/atomicwriter_test.go

```diff
@@ -2,48 +2,232 @@ package atomicwriter
+	t.Run("empty filename", func(t *testing.T) {
+		writer, err := New("", testMode())
+		if writer != nil {
+			t.Errorf("Should not have created writer")
+		}
+		if err == nil || err.Error() != "file name is empty" {
+			t.Errorf("Should produce a 'file name is empty' error, but got %[1]T (%[1]v)", err)
+		}
+	})
+	t.Run("directory", func(t *testing.T) {
+		tmpDir := t.TempDir()
+		writer, err := New(tmpDir, testMode())
+		if writer != nil {
+			t.Errorf("Should not have created writer")
+		}
+		if err == nil || err.Error() != "cannot write to a directory" {
+			t.Errorf("Should produce a 'cannot write to a directory' error, but got %[1]T (%[1]v)", err)
+		}
+	})
+}
+
+func TestWriteFile(t *testing.T) {
+	t.Run("empty filename", func(t *testing.T) {
+		err := WriteFile("", nil, testMode())
+		if err == nil || err.Error() != "file name is empty" {
+			t.Errorf("Should produce a 'file name is empty' error, but got %[1]T (%[1]v)", err)
+		}
+	})
+	t.Run("write to directory", func(t *testing.T) {
+		err := WriteFile(t.TempDir(), nil, testMode())
+		if err == nil || err.Error() != "cannot write to a directory" {
+			t.Errorf("Should produce a 'cannot write to a directory' error, but got %[1]T (%[1]v)", err)
+		}
+	})
+	t.Run("write to file", func(t *testing.T) {
+		tmpDir := t.TempDir()
+		fileName := filepath.Join(tmpDir, "test.txt")
+		fileContent := []byte("file content")
+		fileMode := testMode()
+		if err := WriteFile(fileName, fileContent, fileMode); err != nil {
+			t.Fatalf("Error writing to file: %v", err)
+		}
+		assertFile(t, fileName, fileContent, fileMode)
+		assertFileCount(t, tmpDir, 1)
+	})
+	t.Run("missing parent directory", func(t *testing.T) {
+		tmpDir := t.TempDir()
+		fileName := filepath.Join(tmpDir, "missing-dir", "test.txt")
+		fileContent := []byte("file content")
+		fileMode := testMode()
+		if err := WriteFile(fileName, fileContent, fileMode); !errors.Is(err, os.ErrNotExist) {
+			t.Errorf("Should produce a 'not found' error, but got %[1]T (%[1]v)", err)
+		}
+		assertFileCount(t, tmpDir, 0)
+	})
+	t.Run("symlinked file", func(t *testing.T) {
+		tmpDir := t.TempDir()
+		linkTarget := filepath.Join(tmpDir, "symlink-target")
+		if err := os.WriteFile(linkTarget, []byte("orig content"), testMode()); err != nil {
+			t.Fatal(err)
+		}
+		if err := os.Symlink(linkTarget, filepath.Join(tmpDir, "symlinked-file")); err != nil {
+			t.Fatal(err)
+		}
+		origFileCount := 2
+		assertFileCount(t, tmpDir, origFileCount)
+
+		fileName := filepath.Join(tmpDir, "symlinked-file")
+		fileContent := []byte("new content")
+		fileMode := testMode()
+		if err := WriteFile(fileName, fileContent, fileMode); err != nil {
+			t.Fatalf("Error writing to file: %v", err)
+		}
+		assertFile(t, fileName, fileContent, fileMode)
+		assertFileCount(t, tmpDir, origFileCount)
+		// FIXME(thaJeztah): [os.Rename] does not resolve symlinks, so writing to a symlinked location replaces the link with a file.
+		// assertFile(t, linkTarget, fileContent, fileMode)
```

Comment:
The current situation is that if the destination is a symlink, it gets replaced by a file (because `os.Rename` doesn't follow symlinks.

Reply:
IMO, while it might be inconvenient, it's reasonable approach

Reply:
@vvoland trying to interpret your comment; do you mean the current (don't resolve symlinks) is "inconvenient, but reasonable", or  do you mean; "changing this is inconvenient, but something to do"?

I'm considering the latter (changing the behavior to resolve symlinks), so that these writers can act basically as a drop-in for `os.WriteFile` (and similar).

Reply:
I meant the current behavior.

Looking at our usages of the atomic writer, I don't think there's an actual reason we'd need this behavior.
Overall symlink traversal is a can of worms so I think it's reasonable to try to avoid opening it if possible 🙈 

On the other hand, this might be "convenient" if someone used this package as a dependency in different context (outside moby). I think we should optimize the package design for our usage inside moby though.

Reply:
Yeah, so I started looking into this code because I wanted to swap some code in docker/cli to use the atomic writer package instead of inventing its own;


- https://github.com/docker/cli/pull/5914/files


(CLI also had a check for validating the destination path, which also excluded non-regular files FWIW)